File: pe-core/src/main/java/org/saar/core/camera/projection/ScreenPerspectiveProjection.java
package org.saar.core.camera.projection;

import org.saar.core.camera.Projection;
import org.saar.core.screen.Screen;

public class ScreenPerspectiveProjection extends PerspectiveProjection implements Projection {

    private final Screen screen;
    private float fov;
    private float near;
    private float far;

    public ScreenPerspectiveProjection(Screen screen, float fov, float near, float far) {
        this.screen = screen;
        this.fov = fov;
        this.near = near;
        this.far = far;
    }

    @Override
    public float getFov() {
        return this.fov;
    }

    public void setFov(float fov) {
        this.fov = fov;
    }

    @Override
    public float getWidth() {
        return this.screen.getWidth();
    }

    @Override
    public float getHeight() {
        return this.screen.getHeight();
    }

    @Override
    public float getNear() {
        return this.near;
    }

    public void setNear(float near) {
        this.near = near;
    }

    @Override
    public float getFar() {
        return this.far;
    }

    public void setFar(float far) {
        this.far = far;
    }
}

File: pe-core/src/main/java/org/saar/core/camera/projection/ScreenPerspectiveProjection.kt
package org.saar.core.camera.projection

import org.joml.Matrix4f
import org.saar.core.camera.Projection
import org.saar.core.screen.MainScreen
import org.saar.core.screen.Screen
import org.saar.maths.utils.Matrix4
import org.saar.maths.utils.Matrix4.ofProjection

class ScreenPerspectiveProjection(
    private val screen: Screen = MainScreen,
    override var fov: Float,
    override var near: Float,
    override var far: Float
) : PerspectiveProjection, Projection {

    constructor(fov: Float, near: Float, far: Float) : this(MainScreen, fov, near, far)

    override val width: Float get() = this.screen.width.toFloat()

    override val height: Float get() = this.screen.height.toFloat()

    override val matrix: Matrix4f = Matrix4.create()
        get() = ofProjection(
            Math.toRadians(this.fov.toDouble()).toFloat(),
            this.width, this.height, this.near, this.far, field
        )
}
